WIP: stricter tsc options

winksaville opened the pull request:
Changed tsconfig.json to extend tsconfig.base.json and added
tsconfig.base.json and two version of it ".full" for doing the
conversion and ".min" for allow the projec to continue to pass.

I used "yarn build | less" when doing the conversion.

<!-- Hey, there! Thanks for contributing to Alsatian,
      Add a quick description of the changes you have made below -->

# Description

<!-- Now, we've created a handy checklist before submitting your pull request to ensure it goes smoothly
      (we checked the first one for you because we know it's true) -->

# Checklist

- [x] I am an awesome developer and proud of my code
- [x] I added / updated / removed relevant unit or integration tests to prove my change works
- [x] I ran all tests using ```npm test``` to make sure everything else still works
- [x] I ran ```npm run review``` to ensure the code adheres to the repository standards

<!-- Thanks again, and if you'd like to add any further information you can do so below -->

# Additional Information

winksaville commented:
A first stab at using stricter compiler options.

winksaville commented:
Strawman for #383.

jamesrichford commented on cli/alsatian-cli-options.ts:
```diff
@@ -73,7 +73,7 @@ export class AlsatianCliOptions {
 
          // filter out the timeout argument and its value
-         return args.filter((value, index) => {
+         return args.filter((_, index) => {
```
This is one thing I don't like about the no unused arguments. Granted `filter` is a common function but it's obscuring argument names for no good reason.

winksaville replied:
I disagree, a call back/lambda function is being passed to filter and using "_" clearly states that in the implementation of this function the first parameter isn't used and that it wasn't an oversight.

winksaville replied:
So do you want this to be?:
```
return arc.filter((value, index) => {
   value = null; // Undefined
```

jamesrichford replied:
I'd say not updating the `value` and just leaving it as it is due to it being something we're being passed rather than an interface we're defining as in the other comment if that makes sense. Absolutely up for debate though as it's more of a group decision :)

Jameskmonger replied:
I think leave it as it is. Any odd maskings around `value` (such as reassigning it, using an underscore) could cause confusion.

jamesrichford commented on core/decorators/async-setup-decorator.ts:
```diff
@@ -5,6 +5,7 @@ import { SETUP } from "./_metadata-keys";
 export function AsyncSetup(target: object,
                            decoratedPropertyKey: string,
                            descriptor?: TypedPropertyDescriptor<() => any>) {
+   descriptor = undefined; // Unused
```
Similarly this seems like it's just doing too much I'm not sure I agree with the no unused arguments

winksaville replied:
Depends how strict you want to be, I try to be as strict as possible and having the compiler inform you that a parameter isn't being used I think is a good idea. I do which there was a UNUSED macro like I have in C/C++. I also tried to use `<void>descriptor;` and that satisfied the compiler but the runtime didn't like it :)

jamesrichford replied:
Haha boo on the compiler not liking it!

I guess my thoughts are yes, agree having the warning for unused arguments is a good idea. The only exception I'd say is in these two cases where we're implementing an interface rather than creating our own. If we control the interface absolutely rip out unused arguments but I guess I just like to see what's going on with what we'd be provided if that makes sense? thoughts?

jamesrichford commented on core/errors/error-match-error.ts:
```diff
@@ -2,7 +2,7 @@ import { MatchError } from "./match-error";
 export class ErrorMatchError extends MatchError {
 
-   public constructor(actualError: Error,
+   public constructor(actualError: Error | undefined | null,
```
Should this be the union type everywhere? Perhaps it should be cast instead at some point?

winksaville replied:
I tend to try not to uses casts if there is another way. I'm not sure what you mean by "Should this be the union type everywhere" as the compiler will be insisting this union type will need to be used "everywhere". Reading about [advanced types](https://www.typescriptlang.org/docs/handbook/advanced-types.html) we could create an Interface or a Type Alias. Type Alias seems a possibility, maybe something like one of these or maybe you've got something else in mind:
```
type ErrorOrUndefOrNull = Error | undefined | null;
type ErrorOrUndefNull = Error | undefined | null;
type ErrorUndefNull = Error | undefined | null;
```

jamesrichford replied:
Sure, I'm thinking that it should be Error | null though and keep a consistency here in that it is nothing on purpose i.e. `null`?

winksaville replied:
I had to  add both because both assumptions were made, I'll try for just null. Do you want to use a union type, `Error | null`, or a type alias, say `type NullableError = Error | null`?

jamesrichford replied:
I think I'd be happy with `Error | null` it was more the three unions seemed too much :)

winksaville commented:
I rebased on top of master and fixed some suggestions.

Added type alias `ErrorUndefOrNull` I couldn't just use `Error | null` because the `actualMatch` can be `undefined` or `null`, i.e. we have `toBeUndefined` and `toBeNull`

Removed my use of underscore for the `value` parameter in the filter callbask and instead assign and empty string to `value`  and marked Unused.

Let me know what you think

winksaville commented:
@jamesrichford @Jameskmonger what do you guys think?

jamesrichford commented:
Hey @winksaville,

Just a few questions from me:

1) What is the purpose of setting the unused values to a default value?

2) What is the plan with the base tsconfig.json files?

Also I think I didn't make myself clear about the `Error | null` thing. I meant rather that we should refactor out the case where it can be `undefined` as it's probably better to be setting it definitely to `null` in this case. Thoughts?

Overall I really like it, we should be being as strict as possible on quality so double 👍 👍 (plus two if that's possible haha)

What about you @Jameskmonger?

winksaville commented:
Setting the unused entity to a value satifies the compiler that it was used.

As for tsconfig.base.json, we can keep it or not. I use it to make it easy
to know what the typical settings are when there are multiple tsconfig.json
files. For now I find convenient, butt add I said in the wend we can keep
it out not.

I'm not sure how to refactor it since a single param/variable can take 3
values (a valid value, undefined or null) how to refactor. We can do it
later unless you can educate me.

So I'll continue down this path, please review as I make changes so as to
spread the number of review comments at any one time.

On Tue, May 16, 2017, 10:52 AM James Richford <notifications@github.com>
wrote:

> Hey @winksaville <https://github.com/winksaville>,
>
> Just a few questions from me:
>
>    1.
>
>    What is the purpose of setting the unused values to a default value?
>    2.
>
>    What is the plan with the base tsconfig.json files?
>
> Also I think I didn't make myself clear about the Error | null thing. I
> meant rather that we should refactor out the case where it can be
> undefined as it's probably better to be setting it definitely to null in
> this case. Thoughts?
>
> Overall I really like it, we should be being as strict as possible on
> quality so double 👍 👍 (plus two if that's possible haha)
>
> What about you @Jameskmonger <https://github.com/jameskmonger>?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/alsatian-test/alsatian/pull/390#issuecomment-301861862>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-hHAL7Ix3Y5Kf5VG4-TLwOUOy6udaAks5r6eJIgaJpZM4NXYsP>
> .
>

jamesrichford commented:
@winksaville - is the a reason the error would ever be undefined as opposed to null?

winksaville commented:
Yes, within the code base Error can be undefined, Any time a variable is
not initialized explicitly its undefined. Also anytime a parameter is
optional, i.e :?, it's passed as undefined.

On Wed, May 17, 2017, 1:30 PM James Richford <notifications@github.com>
wrote:

> @winksaville <https://github.com/winksaville> - is the a reason the error
> would ever be undefined as opposed to null?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/alsatian-test/alsatian/pull/390#issuecomment-302222514>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-hHEC5KZcC1a2_AzcY_tkKvSo5gUNCks5r61jQgaJpZM4NXYsP>
> .
>

winksaville commented:
It's probably possible to never have undefined within the code base but I'd
make that a second pass. At least that is my suggestion.

On Wed, May 17, 2017, 7:30 PM Wink Saville <notifications@github.com> wrote:

> Yes, within the code base Error can be undefined, Any time a variable is
> not initialized explicitly its undefined. Also anytime a parameter is
> optional, i.e :?, it's passed as undefined.
>
> On Wed, May 17, 2017, 1:30 PM James Richford <notifications@github.com>
> wrote:
>
> > @winksaville <https://github.com/winksaville> - is the a reason the
> error
> > would ever be undefined as opposed to null?
> >
> > —
> > You are receiving this because you were mentioned.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/alsatian-test/alsatian/pull/390#issuecomment-302222514
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AA-hHEC5KZcC1a2_AzcY_tkKvSo5gUNCks5r61jQgaJpZM4NXYsP
> >
> > .
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alsatian-test/alsatian/pull/390#issuecomment-302283546>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-hHPRC5jU9I9u5wmyQBjY0tWyvGifSks5r660ygaJpZM4NXYsP>
> .
>

winksaville commented:
Despite what I said in my previous post, I've now figured out how to change ErrorUndefOrNull to ErrorOrNull. There is still a ton of work to do, but I think I've made some progress and would like to get in what I've changed.

What do you guys think?

jamesrichford commented:
@winksaville I'll take a look at this asap, hopefully later today :)

winksaville commented:
Take particular note of:

1) Keep tsconfig.base.json (I'd like to but no big deal)
2) Is `Unused` method a acceptable solution for being able to use noUnusedLocals and Parameters.
3) Is `ErrorOrNull` acceptable alternatively we could just using a simple type union everywhere `Error | null`
4) ... :)

jamesrichford commented:
Thanks @winksaville for the breakdown.

I think the only points I want I to address are those

1. What is the purpose of a base.min and full for the tsconfig.json?
2. I have been discussing this with @jameskmonger and the feeling we're getting is that this rule (unused arguments) is good in principle but in the case of implementing an existing interface (e.g. callback) it seems we're doing weird stuff to get this rule to work. Perhaps we need to contact TypeScript about this rule and at exceptions? I do like the solution you've come up with it's quite clear :)
3. I think `Error | null` as we have that pattern elsewhere and it's just as clear.
4. Nice work :)

winksaville commented:
So I've taken care of 1 and 3. For 2, I'm going leave in the `Unused(xxx)` as its not that intrusive, they can always be removed later. As for 4, I'll try to continue doing nice work ;)

Let me know what else you'd like me to do before committing this change, one thing would be to change the branch name, but then we loose. Next time I won't use "WIP"

jamesrichford commented:
@winksaville perfect :) should the `tsconfig.base.json` not have all the new flags you suggested in it?

winksaville commented:
No, because it won't build until "all" files are updated, this is only the first step.

jamesrichford commented:
Ah, I see :). How many other files are there to fix?

winksaville commented:
Many there are about 500+ errors right now see [here](https://gist.github.com/winksaville/1e0ef24ad21fe534d452d5a0684635c7).

jamesrichford commented:
Great looks like mainly specs so I'll just pop a note on #383 :)

Merging! Great work :)
